In [72]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
import statsmodels.formula.api as smf
from IPython.display import display, HTML

import warnings 
warnings.simplefilter("ignore")

Данные к заданиям 1—3 находятся в файле Ch56data.xlsx.
1. На листе Problem 1 перечислено количество автомобилей, каждый день запаркованных как на открытой, так и закрытой автостоянке возле бизнес-школы Kelley School of Business Университета Индианы. Найдите коэффициент корреляции и объясните корреляцию между количеством автомобилей, запаркованных на открытой и закрытой стоянках.
2. На листе Problem 2 указаны ежедневные объемы продаж (в долларах) для лазерных принтеров, картриджей и канцелярских принадлежностей. Найдите коэффициенты корреляции и объясните корреляцию между этими величинами.
3. На листе Problem 3 показаны годовые доходы по акциям, казначейским векселям и облигациям. Найдите коэффициенты корреляции и объясните корреляцию между годовыми доходами по этим трем классам инвестиций.

In [2]:
parking = pd.read_excel('~/Study/excel/Ch56/Practice Files/Ch56data.xlsx', skiprows=2, usecols=range(2,4))

In [3]:
parking

Outdoor  Garage
0      140     180
1      120     200
2      130     190
3      110     210
4      160     160
5      135     185

In [4]:
parking.corr()

Outdoor  Garage
Outdoor      1.0    -1.0
Garage      -1.0     1.0

*Цифры показывают строго отрицательную линейную связь. Когда идет дождь или жарко или холодно, гараж, вероятно, чаще используется. В противном случае люди, скорее всего, предпочтут парковаться на улице.*

In [5]:
laser = pd.read_excel('~/Study/excel/Ch56/Practice Files/Ch56data.xlsx',
                      skiprows=1, usecols=range(2,5), sheet_name='Problem 2')

In [7]:
laser.head()

Laser Printers  Cartridges  School Supplies
0           17291        4379             3618
1           13734        2258             3514
2           18802        4206             3587
3           12171        2137             3007
4           16402        3775             3850

In [8]:
laser.corr()

Laser Printers  Cartridges  School Supplies
Laser Printers         1.000000    0.859024         0.167991
Cartridges             0.859024    1.000000         0.142642
School Supplies        0.167991    0.142642         1.000000

*Когда спрос на принтеры высок, он также высок и для картриджей. Когда спрос на школьные принадлежности высок,
это мало что говорит о спросе на принтеры или картриджи.*

In [14]:
stocks = pd.read_excel('~/Study/excel/Ch56/Practice Files/Ch56data.xlsx',
                      skiprows=6, usecols=range(4), sheet_name='Problem 3')

In [17]:
stocks.head()

Year  Stocks  T.Bills  T.Bonds
0  1928  0.4381   0.0308   0.0084
1  1929 -0.0830   0.0316   0.0420
2  1930 -0.2512   0.0455   0.0454
3  1931 -0.4384   0.0231  -0.0256
4  1932 -0.0864   0.0107   0.0879

In [20]:
stocks.loc[:,'Stocks': 'T.Bonds'].corr()

Stocks   T.Bills   T.Bonds
Stocks   1.000000 -0.001443  0.123372
T.Bills -0.001443  1.000000  0.382041
T.Bonds  0.123372  0.382041  1.000000

*Знания о доходности казначейских векселей мало что нам говорят о доходности акций. Векселя и облигации демонстрируют умеренную положительную корреляцию так что, если доходность векселей растет, это немного увеличивает вероятность того, что облигации будут приносить больше. Это потому, что доходность как по облигациям, так и по векселям связаны с изменением процентных ставок. Акции и облигации имеют небольшую положительную взаимосвязь.*

***
В файле Dow.xlsx находятся ежемесячные данные о доходах по 30 акциям, входящим в индекс Доу—Джонса. Найдите коэффициенты корреляции для всех акций. Затем для каждой акции с помощью условного форматирования выделите три акции, наиболее коррелированные с этой акцией. (Безусловно, корреляция акции с ней самой не должна быть выделена.)

In [21]:
dow = pd.read_excel('~/Study/excel/Ch56/Practice Files/Dow.xlsx', skiprows=2).dropna()

In [22]:
dow.head()

Date        AA       AIG       AXP        BA         C       CAT  \
1 1986-08-01  0.096296  0.050093  0.098028  0.029596  0.096273  0.117634   
2 1986-09-02  0.000000 -0.128975 -0.140898 -0.123193 -0.075127 -0.240577   
3 1986-10-01 -0.016757 -0.008114  0.042380 -0.009367  0.073511  0.052798   
4 1986-11-03 -0.048103 -0.509202 -0.019215 -0.035552 -0.031386 -0.006269   
5 1986-12-01 -0.021658  0.018833 -0.009370  0.002549 -0.008837  0.012617   

         DD       DIS        GE  ...        MO       MRK      MSFT       PFE  \
1  0.083816 -0.056842  0.076986  ...  0.055556  0.085634  0.000000  0.018727   
2 -0.027316 -0.134341 -0.085856  ... -0.128289 -0.148071 -0.008772 -0.147059   
3  0.073390  0.110886  0.060943  ...  0.109434  0.102741  0.371681  0.030172   
4  0.039186 -0.017092  0.088525  ...  0.010204  0.017308  0.283871  0.020921   
5 -0.060087  0.000000  0.036145  ... -0.032054  0.120995 -0.030151  0.000000   

         PG         T       UTX        VZ       WMT       XOM  
1  0.033993  0.108804  0.115487  0.027211  0.010753  0.133663  
2 -0.153125 -0.085182 -0.079119 -0.147417 -0.106383 -0.023523  
3  0.107011  0.011792  0.014911  0.038838  0.068571  0.011152  
4  0.036667  0.043077  0.040812  0.033647  0.061275  0.018382  
5 -0.017749  0.003397  0.030697 -0.023438 -0.023725  0.012563  

[5 rows x 31 columns]

In [28]:
def highlight_top_corr(s, n=3):
    is_top = s.isin(s.nlargest(n+1)[1:])
    return ['background-color: yellow' if v else '' for v in is_top]

In [47]:
def plot_pandas_style(styler):
    html = '\n'.join([line.lstrip() for line in styler.render().split('\n')])
    return HTML(html)

In [48]:
html = (dow.iloc[:, 1:].corr().style \
               .apply(highlight_top_corr, axis=1)
        )
plot_pandas_style(html)

***
Команды НФЛ проводят в регулярном сезоне 16 игр. Предположим, что стандартное отклонение от количества выигранных игр для всех команд равно 2, и коэффициент корреляции между количеством выигранных командой игр в двух подряд идущих сезонах равен 0,5. Если команда идет в регулярном сезоне со счетом 12:4, каков наилучший прогноз для количества игр, которые команда выиграет в следующем сезоне?

*В этом году мы выиграли на 2 стандартных отклонения больше среднего (в среднем 8). Поэтому в следующем году мы прогнозируем, что выиграем на 0,5 (2) = 1 стандартное отклонение больше среднего, что составляет 10 игр.*

***
Фицци Драгс (Fizzy Drugs) должен оптимизировать продуктивность важного химического процесса. Компания считает, что количество произведенной продукции в фунтах при каждом запуске процесса зависит от размера используемого резервуара, давления и температуры. Привлеченные к решению задачи ученые считают, что результат изменения одной переменной может зависеть от значений других переменных. Резервуар должен иметь объем от 1,3 до 1,5 м3, кроме того, необходимо поддерживать давление между 4 и 4,5 мм рт. ст. и температуру между 22 и 30 градусами Цельсия. Ученые провели эксперименты для нижних и верхних уровней трех контрольных переменных, результаты которых представлены в файле Fizzy.xlsx.
1. Определите взаимосвязь между количеством продукции, объемом резервуара, температурой и давлением.
2. Рассмотрите взаимосвязь между давлением, объемом резервуара и температурой.
3. Какие значения температуры и давления и какой объем резервуара следует порекомендовать?

In [6]:
fizzy = pd.read_excel('~/Study/excel/Ch59/Practice Files/Fizzy.xlsx', skiprows=1)

In [19]:
fizzy.rename(columns={'S*P': 'SP', 'S*T': 'ST', 'P*T': 'PT'}, inplace=True)

In [20]:
model = smf.ols("Yield ~ Size + Pressure + Temperature + SP + ST + PT", data=fizzy).fit()

In [24]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Yield   R-squared:                       0.999
Model:                            OLS   Adj. R-squared:                  0.998
Method:                 Least Squares   F-statistic:                     1063.
Date:                Mon, 31 May 2021   Prob (F-statistic):           2.63e-12
Time:                        21:18:34   Log-Likelihood:                -61.030
No. Observations:                  16   AIC:                             136.1
Df Residuals:                       9   BIC:                             141.5
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept   -1.888e+04   1019.339    -18.525      0.000   -2.12e+04   -1.66e+04
Size         1.197e+04    666.719     17.959      0.000    1.05e+04    1.35e+04
Pressure     3825.6250    226.306     16.905      0.000    3313.685    4337.565
Temperature   111.0938     20.159      5.511      0.000      65.491     156.697
SP          -2012.5000    146.309    -13.755      0.000   -2343.473   -1681.527
ST            -89.8438      9.144     -9.825      0.000    -110.530     -69.158
PT              1.5625      3.658      0.427      0.679      -6.712       9.837
==============================================================================
Omnibus:                        0.280   Durbin-Watson:                   1.827
Prob(Omnibus):                  0.870   Jarque-Bera (JB):                0.444
Skew:                           0.045   Prob(JB):                        0.801
Kurtosis:                       2.189   Cond. No.                     4.04e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.04e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Все факторы значимы, за исключением `P*T`, поэтому мы отбрасываем этот столбец и запустим модель снова.

In [25]:
model = smf.ols("Yield ~ Size + Pressure + Temperature + SP + ST - PT", data=fizzy).fit()

In [26]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Yield   R-squared:                       0.999
Model:                            OLS   Adj. R-squared:                  0.998
Method:                 Least Squares   F-statistic:                     1390.
Date:                Mon, 31 May 2021   Prob (F-statistic):           7.18e-14
Time:                        21:21:27   Log-Likelihood:                -61.191
No. Observations:                  16   AIC:                             134.4
Df Residuals:                      10   BIC:                             139.0
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept   -1.906e+04    896.718    -21.251      0.000   -2.11e+04   -1.71e+04
Size         1.197e+04    638.885     18.741      0.000    1.05e+04    1.34e+04
Pressure     3866.2500    196.781     19.647      0.000    3427.794    4304.706
Temperature   117.7344     12.299      9.573      0.000      90.331     145.138
SP          -2012.5000    140.201    -14.354      0.000   -2324.887   -1700.113
ST            -89.8438      8.763    -10.253      0.000    -109.368     -70.320
==============================================================================
Omnibus:                        0.640   Durbin-Watson:                   1.811
Prob(Omnibus):                  0.726   Jarque-Bera (JB):                0.611
Skew:                           0.100   Prob(JB):                        0.737
Kurtosis:                       2.064   Cond. No.                     1.47e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.47e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Yield = -19056 + 11973 * Size + 3866 * Pressure + 118 * TEMP - 2012 SP - 90 * ST

1. По мере увеличения размера эффект увеличения давления уменьшается.
2. По мере увеличения размера эффект увеличения температуры уменьшается. Все значения `p` значимы.
3. Производство продукциии максимально при температуре и давлении самого большого значения.

***
В супермаркете Mr. D’s в течение 12 недель подряд проводились наблюдения за продажей (учитывалось количество продаж) консервированных томатов. (См. файл Grocery.xlsx.) Каждую неделю отслеживались следующие данные.
- Во все ли тележки были положены рекламные листовки о консервированных томатах?
- Каждому ли покупателю были вручены купоны на консервированные томаты?
- Была ли снижена цена (нет, на 1 или 2 цента)?

На основе этих данных определите, как перечисленные факторы влияют на продажи. Предскажите продажи консервированных томатов для недели, на которой в тележки были положены рекламные листовки, покупателям предлагались купоны, и цена на томаты была снижена на 1 цент.

In [28]:
grocery = pd.read_excel('~/Study/excel/Ch59/Practice Files/Grocery.xlsx', skiprows=3)

In [50]:
grocery.rename(columns={'Cart?': 'Cart', 'Coupon??': 'Coupon'}, inplace=True)

In [43]:
grocery['Cart?'] = grocery['Cart Notice?'].map({'Yes': 1, 'No': 0})
grocery['Coupon??'] = grocery['Coupon?'].map({'Yes': 1, 'No': 0})

In [46]:
grocery = pd.concat([grocery, pd.get_dummies(grocery['Price Reduction?'].astype(str), 
                                            drop_first=True, prefix='cents_off')], axis=1)

In [52]:
grocery

Week Cart Notice? Coupon?  Price Reduction?  Sales  Cart  Coupon  \
0      1          Yes     Yes                 0     36     1       1   
1      2          Yes     Yes                 1     38     1       1   
2      3          Yes     Yes                 2     40     1       1   
3      4          Yes      No                 0     40     1       0   
4      5          Yes      No                 1     42     1       0   
5      6          Yes      No                 2     44     1       0   
6      7           No     Yes                 0     12     0       1   
7      8           No     Yes                 1     20     0       1   
8      9           No     Yes                 2     30     0       1   
9     10           No      No                 0      8     0       0   
10    11           No      No                 1     16     0       0   
11    12           No      No                 2     33     0       0   

    cents_off_1  cents_off_2  
0             0            0  
1             1            0  
2             0            1  
3             0            0  
4             1            0  
5             0            1  
6             0            0  
7             1            0  
8             0            1  
9             0            0  
10            1            0  
11            0            1

In [53]:
model = smf.ols("Sales ~ Cart + Coupon + cents_off_1 + cents_off_2", data=grocery).fit()

In [56]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.887
Model:                            OLS   Adj. R-squared:                  0.822
Method:                 Least Squares   F-statistic:                     13.70
Date:                Mon, 31 May 2021   Prob (F-statistic):            0.00201
Time:                        22:41:40   Log-Likelihood:                -33.865
No. Observations:                  12   AIC:                             77.73
Df Residuals:                       7   BIC:                             80.15
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      14.5000      3.438      4.217      0.004       6.370      22.630
Cart           20.1667      3.075      6.558      0.000      12.895      27.438
Coupon         -1.1667      3.075     -0.379      0.716      -8.438       6.105
cents_off_1     5.0000      3.766      1.328      0.226      -3.906      13.906
cents_off_2    12.7500      3.766      3.385      0.012       3.844      21.656
==============================================================================
Omnibus:                        1.544   Durbin-Watson:                   2.215
Prob(Omnibus):                  0.462   Jarque-Bera (JB):                0.845
Skew:                          -0.209   Prob(JB):                        0.655
Kurtosis:                       1.769   Cond. No.                         4.67
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Скидка в 1 цент и купон не имеют значения, поэтому удалим их.

In [57]:
model = smf.ols("Sales ~ Cart + cents_off_2", data=grocery).fit()

In [58]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.856
Model:                            OLS   Adj. R-squared:                  0.824
Method:                 Least Squares   F-statistic:                     26.72
Date:                Mon, 31 May 2021   Prob (F-statistic):           0.000164
Time:                        22:44:17   Log-Likelihood:                -35.310
No. Observations:                  12   AIC:                             76.62
Df Residuals:                       9   BIC:                             78.08
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      16.4167      2.418      6.788      0.000      10.946      21.888
Cart           20.1667      3.059      6.592      0.000      13.247      27.087
cents_off_2    10.2500      3.245      3.159      0.012       2.910      17.590
==============================================================================
Omnibus:                        1.123   Durbin-Watson:                   2.230
Prob(Omnibus):                  0.570   Jarque-Bera (JB):                0.870
Skew:                          -0.419   Prob(JB):                        0.647
Kurtosis:                       1.982   Cond. No.                         2.93
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Листовка в корзине и скидка в 2 цента имеют большое значение. Наличие уведомления в тележке увеличивает продажи на 20. Скидка 2 цента увеличивает продажи на 10.

Прогноз для «тележки были положены рекламные листовки, покупателям предлагались купоны, и цена на томаты была снижена на 1 цент»: 16,41 + 20,17 = 36,58.

***
В файле Countryregion.xlsx для нескольких слаборазвитых стран содержатся следующие данные:
- детская смертность;
- уровень грамотности взрослого населения;
- процент учеников, закончивших начальную школу;
- ВНП на душу населения.
Исходя из этих данных, создайте уравнение для прогнозирования детской смертности. Имеются ли в этом наборе данных выбросы? Проинтерпретируйте коэффициенты полученного уравнения. В пределах каких значений 95% прогнозов по детской смертности должны быть точными?

In [65]:
сountries = pd.read_excel('~/Study/excel/Ch59/Practice Files/Countryregion.xlsx',
                          usecols=range(1, 6), skiprows=3).rename(columns={'Unnamed: 1': 'Country'})

In [70]:
сountries.columns = сountries.columns.str.replace(r'\(.*\)', '', regex=True)

In [71]:
сountries

Country  Infant Mortality  age adult literacy  \
0           Cuba                18                  98   
1      Sri Lanka                20                  85   
2     Costa Rica                19                  94   
3        Vietnam                44                  85   
4          China                54                  80   
5   South Africa                56                  76   
6   Saudi Arabia                38                  59   
7         Brazil                60                  78   
8        Zimbawe                68                  82   
9        Morocco                68                  42   
10      Pakistan                98                  36   
11       Nigeria                86                  44   

    age finishing primary school  GNP per capita  
0                             98            2000  
1                             92            3300  
2                             84            5800  
3                             58             600  
4                             86            2400  
5                             68            4000  
6                             68           11000  
7                             56            5600  
8                             76            1800  
9                             76            3400  
10                            38            2100  
11                            56            1600

In [78]:
model = smf.ols("Q('Infant Mortality') ~ Q('age adult literacy') \
+ Q('age finishing primary school') + Q('GNP per capita')", data=сountries).fit()

In [79]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                              OLS Regression Results                             
=================================================================================
Dep. Variable:     Q('Infant Mortality')   R-squared:                       0.846
Model:                               OLS   Adj. R-squared:                  0.788
Method:                    Least Squares   F-statistic:                     14.64
Date:                   Mon, 31 May 2021   Prob (F-statistic):            0.00130
Time:                           23:45:24   Log-Likelihood:                -44.347
No. Observations:                     12   AIC:                             96.69
Df Residuals:                          8   BIC:                             98.63
Df Model:                              3                                         
Covariance Type:               nonrobust                                         
=====================================================================================================
                                        coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------
Intercept                           154.2055     16.152      9.547      0.000     116.958     191.452
Q('age adult literacy')              -0.6772      0.229     -2.961      0.018      -1.205      -0.150
Q('age finishing primary school')    -0.5982      0.276     -2.169      0.062      -1.234       0.038
Q('GNP per capita')                  -0.0029      0.001     -2.271      0.053      -0.006    4.52e-05
==============================================================================
Omnibus:                        0.686   Durbin-Watson:                   1.732
Prob(Omnibus):                  0.710   Jarque-Bera (JB):                0.409
Skew:                           0.409   Prob(JB):                        0.815
Kurtosis:                       2.613   Cond. No.                     2.12e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.12e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [83]:
coeffs = model.params.values

In [89]:
predictions = (coeffs[0] + сountries.iloc[:, 2] * coeffs[1]
               + сountries.iloc[:, 3] * coeffs[2] + сountries.iloc[:, 4] * coeffs[3])

In [93]:
сountries['predictions'] = predictions
сountries['residuals'] = сountries['Infant Mortality'] - сountries['predictions'] # model.resid
сountries['residuals^2']  = сountries['residuals'] ** 2 

In [118]:
np.sqrt(model.scale) # Standard Error

11.933992311331808

In [94]:
сountries

Country  Infant Mortality  age adult literacy  \
0           Cuba                18                  98   
1      Sri Lanka                20                  85   
2     Costa Rica                19                  94   
3        Vietnam                44                  85   
4          China                54                  80   
5   South Africa                56                  76   
6   Saudi Arabia                38                  59   
7         Brazil                60                  78   
8        Zimbawe                68                  82   
9        Morocco                68                  42   
10      Pakistan                98                  36   
11       Nigeria                86                  44   

    age finishing primary school  GNP per capita  predictions  residuals  \
0                             98            2000    23.359512  -5.359512   
1                             92            3300    31.944569 -11.944569   
2                             84            5800    23.313134  -4.313134   
3                             58             600    60.191428 -16.191428   
4                             86            2400    41.555745  12.444255   
5                             68            4000    50.345801   5.654199   
6                             68           11000    41.355546  -3.355546   
7                             56            5600    51.483524   8.516476   
8                             76            1800    47.940709  20.059291   
9                             76            3400    70.342024  -2.342024   
10                            38            2100   100.944321  -2.944321   
11                            56            1600    86.223687  -0.223687   

    residuals^2  
0     28.724365  
1    142.672736  
2     18.603128  
3    262.162342  
4    154.859477  
5     31.969970  
6     11.259689  
7     72.530365  
8    402.375168  
9      5.485078  
10     8.669026  
11     0.050036

Для альфа = 0,10 все три переменные значимы, поэтому я бы предсказал: MORT = 154,2 -0,68 * AL -0,60 * FS -0,0029 * GNP

Нет выбросов, потому что нет остатков > 23,87 (standard error ^ 2) по абсолютному значению.

При прочих равных (ceteris paribus), увеличение ВНП на 1000 долларов снижает количество смертей / 1000 рождений на 2,9. Повышение грамотности взрослого населения на 1% снижает смертность на 1000 рождений на 0,67. Увеличение на 1% детей, заканчивающих школу, снижает смертность на 1000 рождений на 0,6.

95% всех прогнозов с точностью до 2 standard error = 23,87 смертей на тысячу.

***
В файле Baseball96.xlsx содержатся следующие данные: засчитанные пробежки, одинарные, двойные, тройные удары, хоум-раны, кражи базы и уолки для каждой команды Главной лиги бейсбола в сезоне 1996 г. На основе этих данных определите влияние одиночных и двойных ударов, а также других действий на пробежки.

In [119]:
baseball = pd.read_excel('~/Study/excel/Ch59/Practice Files/Baseball96.xlsx', skiprows=1)

In [121]:
baseball.head()

Team  Runs  Singles  Doubles  Triples  HR's  BB's  SB's
0    Cleveland   840      952      279       23   207   542   132
1  Chicago Sox   755      982      252       37   146   576   110
2       Boston   791      907      286       31   175   560    99
3    Minnesota   703      974      270       34   120   471   105
4   California   801      927      252       25   186   564    58

In [123]:
baseball.iloc[:, 2:].corr()

Singles   Doubles   Triples      HR's      BB's      SB's
Singles  1.000000  0.208208  0.000673 -0.184959  0.157080  0.220060
Doubles  0.208208  1.000000  0.092167  0.115713  0.212787  0.082975
Triples  0.000673  0.092167  1.000000 -0.001340 -0.127301  0.017332
HR's    -0.184959  0.115713 -0.001340  1.000000  0.394244  0.098529
BB's     0.157080  0.212787 -0.127301  0.394244  1.000000 -0.087443
SB's     0.220060  0.082975  0.017332  0.098529 -0.087443  1.000000

In [129]:
model = smf.ols("""Runs ~ Singles + Doubles + Triples + Q("HR's") + Q("BB's") + Q("SB's")""", data=baseball).fit()

In [130]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   Runs   R-squared:                       0.942
Model:                            OLS   Adj. R-squared:                  0.926
Method:                 Least Squares   F-statistic:                     56.97
Date:                Tue, 01 Jun 2021   Prob (F-statistic):           6.55e-12
Time:                        00:33:24   Log-Likelihood:                -117.83
No. Observations:                  28   AIC:                             249.7
Df Residuals:                      21   BIC:                             259.0
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -396.9200     70.951     -5.594      0.000    -544.471    -249.369
Singles        0.5185      0.072      7.203      0.000       0.369       0.668
Doubles        0.7829      0.156      5.032      0.000       0.459       1.107
Triples        0.7800      0.559      1.396      0.177      -0.382       1.942
Q("HR's")      1.4060      0.143      9.800      0.000       1.108       1.704
Q("BB's")      0.3750      0.079      4.751      0.000       0.211       0.539
Q("SB's")      0.1635      0.116      1.409      0.173      -0.078       0.405
==============================================================================
Omnibus:                        7.582   Durbin-Watson:                   1.330
Prob(Omnibus):                  0.023   Jarque-Bera (JB):                8.898
Skew:                           0.453   Prob(JB):                       0.0117
Kurtosis:                       5.609   Cond. No.                     2.15e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.15e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [134]:
pd.DataFrame({'Predicted Runs': model.predict(), 'Residuals': model.resid})

Predicted Runs  Residuals
0       848.919480  -8.919480
1       777.643580 -22.643580
2       793.672267  -2.672267
3       708.500953  -5.500953
4       783.006265  17.993735
5       744.569308   4.430692
6       780.630148  15.369852
7       687.989047  52.010953
8       685.213783   5.786217
9       720.205867   9.794133
10      711.443514  -7.443514
11      640.119803 -11.119803
12      682.568353 -40.568353
13      654.881807  -0.881807
14      791.237585  -6.237585
15      742.778804   4.221196
16      657.602651  10.397349
17      741.564100   5.435900
18      643.243304  13.756696
19      688.080879   4.919121
20      657.790367 -23.790367
21      633.998930 -18.998930
22      672.966043   0.033957
23      634.776276  -5.776276
24      611.212676   9.787324
25      657.941439  -5.941439
26      644.924438   0.075562
27      556.518332   6.481668

Мы видим, что сингл стоит 0,52 пробежки; двойной удар стоит 0,78 пробежки, тройной — 0,78 пробежки, хоум-ран — 1,41 пробежки. Кражи базы стоит 0,37 пробежки, уолки — 0,16.

***
В файле Cardata.xlsx приведена следующая информация для 392 моделей автомобилей:
- количество цилиндров;
- объем двигателя;
- мощность в лошадиных силах;
- вес;
- разгон;
- количество миль на галлон (горючего).
Определите уравнение для прогнозирования количества миль на галлон горючего. Почему все независимые переменные не являются значимыми?